### PR TITLE
sql: RETAIN HISTORY for materialized views

### PR DESCRIPTION
**Pull request by mjibson**
Add RETAIN HISTORY to materialized views.

See https://github.com/MaterializeInc/materialize/issues/16701

### Motivation

  * This PR adds a known-desirable feature.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Add `COMPACTION WINDOW` to `MATERIALIZED VIEWS`.

### CODE REVIEW

**Comment by frankmcsherry** [comment]
Is this a good place to bikeshed the name? I think users may not understand "compaction" but might understand e.g. "history" or "historical" windows instead?

File: src/environmentd/tests/sql.rs
Comment (jkosh44) [inline src/environmentd/tests/sql.rs:3579]: Should this be `>=`?

**Review by jkosh44 — APPROVED** [review]
Code LGTM, ~~I don't really have a strong opinion on the naming.~~

EDIT: I forgot I'm on the SQL council this month. I will form an opinion next week.

**Review by benesch — DISMISSED** [review]
Requesting changes since there's a pending discussion on the naming of the option in Slack.

**Comment by mjibson** [comment]
Changed to `RETAIN HISTORY = FOR <interval>`.

File: src/sql-parser/src/parser.rs
Comment (benesch) [inline src/sql-parser/src/parser.rs:3252]: `=` should be optional for consistency with other options (cf `parse_kafka_broker`).

```suggestion
        let _ = self.consume_token(&Token::Eq);
```

File: src/sql/src/plan.rs
Comment (benesch) [inline src/sql/src/plan.rs:1559]: Is this meant to be constructible at the moment? It doesn't seem to be! Is the idea that we'll introduce `RETAIN HISTORY = FOREVER` in the future?

**Review by benesch — COMMENTED** [review]
Syntax looks good modulo the point about the equals sign being optional. 

Just to confirm: it's not possible with this change to _alter_ the compaction window for a materialized view, yeah? Seems fine, since this is plenty to unblock testing with rETL tools.

File: src/sql/src/plan.rs
Comment (mjibson) [reply src/sql/src/plan.rs:1559]: It is written this way to have compatibility with index options which allow for this: https://github.com/MaterializeInc/materialize/blob/a75510b3786759a7e438c1e0f860922f4603002b/src/sql/src/plan.rs#L1548

In a later PR I would like to refactor the index planning and controller read policy code to use this new enum instead of passing around `Option<milliseconds or Timestamp or Duration>`.

**Comment by mjibson** [comment]
> it's not possible with this change to alter the compaction window for a materialized view, yeah?

Correct. ALTER is a bit harder because it has to muck with the SQL. I'd like to ship retain history for CREATE tables, sources and mvs. Then do ALTER afterward.

**Comment by mjibson** [comment]
Hold off on any more reviews, this doesn't work on restart.

File: src/sql/src/plan.rs
Comment (mjibson) [reply src/sql/src/plan.rs:1559]: The principal motivation for this struct is to describe exactly what a `None` Option means. There are some places it means "I want NO user compaction performed" and others it means "no one specified anything, use the default". With the current scheme we'd have some `Option<Option<Duration>>` in a few places. I thought this would improve that greatly, although the refactor is getting annoying.

**Comment by mjibson** [comment]
RFAL. First commit is https://github.com/MaterializeInc/materialize/pull/23854. This now works on restart and has a test for that.

**Review by benesch — COMMENTED** [review]
Syntax LGTM. Not going to have time to look through the rest.

**Comment by ggevay** [comment]
Oh, great! I'm happy about this, because I'm planning to also increase the compaction window of `REFRESH EVERY` materialized views to the refresh interval, and I guess that should be easy after this PR.

I have a question: What happens with indexes on a materialized view that has a custom compaction window? Does the compaction window of the index inherit the MV's window? If not, then this might be surprising to users: they might expect that creating an index can only speed up querying from the MV but never alter the semantics.

File: src/catalog/src/memory/objects.rs
Comment (jkosh44) [inline src/catalog/src/memory/objects.rs:449]: For these objects, what's the difference between `None` and `Some(CustomCompactionWindow::Default)`?

**Review by jkosh44 — APPROVED** [review]

File: src/catalog/src/memory/objects.rs
Comment (mjibson) [reply src/catalog/src/memory/objects.rs:449]: none means the user did not specify a custom window in the syntax. maybe customcompactionwindow::default should not exist then since it's implied by none!

File: test/restart/mzcompose.py
Comment (mjibson) [inline test/restart/mzcompose.py:35]: Putting this test in `restart` doesn't feel quite right. Is there a better place to put it?

**Comment by mjibson** [comment]
The restart test is failing in another test in the file because 

```
$ docker volume rm --force restart_mzdata
--
  | Error response from daemon: remove restart_mzdata: volume is in use - [76a26dde16d3d6561f4829e78674f07de904cbc6a5c330e8cbb2c643b076f28d]
```

Retrying didn't fix it and I'm unable to reproduce locally. Any ideas?

**Comment by nrainer-materialize** [comment]
> The restart test is failing in another test in the file because
> 
> ```
> $ docker volume rm --force restart_mzdata
> --
>   | Error response from daemon: remove restart_mzdata: volume is in use - [76a26dde16d3d6561f4829e78674f07de904cbc6a5c330e8cbb2c643b076f28d]
> ```
> 
> Retrying didn't fix it and I'm unable to reproduce locally. Any ideas?

The full stack trace is:
```
Error response from daemon: remove restart_mzdata: volume is in use - [76a26dde16d3d6561f4829e78674f07de904cbc6a5c330e8cbb2c643b076f28d]
--- mzcompose: test case stash failed: <class 'subprocess.CalledProcessError'>: Command '['docker', 'volume', 'rm', '--force', 'restart_mzdata']' returned non-zero exit status 1.
Traceback (most recent call last):
  File "/var/lib/buildkite-agent/builds/buildkite-15f2293-i-01becf557c126887b-1/materialize/tests/misc/python/materialize/mzcompose/composition.py", line 475, in test_case
    yield
  File "/var/lib/buildkite-agent/builds/buildkite-15f2293-i-01becf557c126887b-1/materialize/tests/test/restart/mzcompose.py", line 433, in workflow_default
    c.workflow(name)
  File "/var/lib/buildkite-agent/builds/buildkite-15f2293-i-01becf557c126887b-1/materialize/tests/misc/python/materialize/mzcompose/composition.py", line 371, in workflow
    func(self)
  File "/var/lib/buildkite-agent/builds/buildkite-15f2293-i-01becf557c126887b-1/materialize/tests/test/restart/mzcompose.py", line 185, in workflow_stash
    c.rm_volumes("mzdata", force=True)
  File "/var/lib/buildkite-agent/builds/buildkite-15f2293-i-01becf557c126887b-1/materialize/tests/misc/python/materialize/mzcompose/composition.py", line 939, in rm_volumes
    spawn.runv(
  File "/var/lib/buildkite-agent/builds/buildkite-15f2293-i-01becf557c126887b-1/materialize/tests/misc/python/materialize/spawn.py", line 72, in runv
    return subprocess.run(
```

[test/restart/mzcompose.py:433](https://github.com/MaterializeInc/materialize/blob/86ee2ca9bf2981b4840b524c83a334b1ffc36d1c/test/restart/mzcompose.py#L433C1-L433C29) just triggers the workflow.
[test/restart/mzcompose.py:185](https://github.com/MaterializeInc/materialize/blob/86ee2ca9bf2981b4840b524c83a334b1ffc36d1c/test/restart/mzcompose.py#L185C15-L185C15) is `c.rm_volumes("mzdata", force=True)` in `workflow_stash`.

@philip-stoev, have you seen this before? Any ideas?

**Comment by mjibson** [comment]
> What happens with indexes on a materialized view that has a custom compaction window? Does the compaction window of the index inherit the MV's window?

Good point! This is a thing we decided on in the read holds design https://github.com/MaterializeInc/materialize/blob/633db6eb8d13bda983c1c9e1ccc812f9961c3314/doc/developer/design/20231102_read_holds.md#L39

And it needs to be tested and probably implemented here.

**Comment by mjibson** [comment]
The docker ps -a log shows `restart-testdrive_no_reset-1` is still running. Could that be related?

**Comment by mjibson** [comment]
> What happens with indexes on a materialized view that has a custom compaction window?

Did some testing and adding an index seems to disable the compaction. Working on a fix.

**Comment by mjibson** [comment]
We decided that indexes breaking this is in spec for now. I'm going to merge this once I figure out the test failure.

**Comment by mjibson** [comment]
@benesch Merging is blocked until your change request is resolved.

**Comment by benesch** [comment]
FYI you can dismiss stale reviews here! 
image

File: test/restart/mzcompose.py
Comment (benesch) [reply test/restart/mzcompose.py:35]: `restart` seems right to me, actually! In the future #team-testing will get you a very fast answer to these sorts of questions. They'd probably recommend adding a platform check for this, since platform checks get interspersed with all sorts of disruptions, like restarts. Can do that as a followup though. cc @MaterializeInc/testing for thoughts.

**Comment by mjibson** [comment]
TIL! Thanks.

**Pull request merged**
